### PR TITLE
Ruff fixes

### PR DESCRIPTION
**Pull request by Fripe070**
This PR should make only 9 linter issues remain. 

The only check still failing is [`N805`](https://docs.astral.sh/ruff/rules/invalid-first-argument-name-for-method/) which will be handled in a separate PR

### CODE REVIEW

**Comment by Sophon96** [comment]
For N805, is it possible to annotate those methods with `@staticmethod`?

**Comment by Fripe070** [comment]
@ThatOtherAndrew I'd like some opinions on the remaining [`PLR0913`](https://docs.astral.sh/ruff/rules/too-many-arguments/) and [`PLR2004`](https://docs.astral.sh/ruff/rules/magic-value-comparison/). Should these rules be kept on and dealt with, or should they be disabled?

**Comment by m31-galaxy** [comment]
> @ThatOtherAndrew I'd like some opinions on the remaining [`PLR0913`](https://docs.astral.sh/ruff/rules/too-many-arguments/) and [`PLR2004`](https://docs.astral.sh/ruff/rules/magic-value-comparison/). Should these rules be kept on and dealt with, or should they be disabled?

Not happy with PLR0913 personally. I feel that on both flagged occasions here, the use of >5 arguments is justified - and I've done my best to make it palatable by using keyword-only and optional arguments. Extracting elsewhere would be much more cumbersome to work with, and there are plenty of stdlib examples with built-in functions accepting more than 5 arguments.

PLR2004 I thought of as a definite no-no in my book, but now that I really consider it, maybe it would actually be good to have? It's pretty much an enforced comment, and worst case scenario is just stating the obvious.

**Comment by Fripe070** [comment]
For PLR0913 would you prefer for it to be disabled outright or just have its limit increased?

And less so a comment and more so a constant if I’ve understood PLR2004 correctly.

**Comment by m31-galaxy** [comment]
I think it's best to just disable - unlike with e.g. line length, I believe that a hard limit does not make sense with function arguments, with what's acceptable depending much more on context and use case than a fixed number. Setting a hard limit will either be low enough to be an inconvenience for legitimate use cases, or high enough that poorly constructed functions will slip right through.

And yes, it is a constant, however what I'm saying is that the constant name will fulfill the role of what'd otherwise most likely be done by a comment. I can see an argument for leaving that enabled

**Comment by Fripe070** [comment]
PR description has been edited to specify that [`N805`](https://docs.astral.sh/ruff/rules/invalid-first-argument-name-for-method/) will be handled in another PR

**Comment by Fripe070** [comment]
I'd like some more feedback on how a fix for [`PLR2004`](https://docs.astral.sh/ruff/rules/magic-value-comparison/) should be implemented.
Should A new file containing constants be created, or should variables/comments be added just where the value is used?

**Comment by Fripe070** [comment]
[`PLR2004`](https://docs.astral.sh/ruff/rules/magic-value-comparison/) is now resolved

File: breadcord/core_modules/module_manager/__init__.py [hunk 1]
```diff
@@ -86,7 +87,7 @@ async def install(self, interaction: discord.Interaction, module: str, branch: s
             f'https://api.github.com/repos/{module}/contents/manifest.toml' + (f'?ref={branch}' if branch else ''),
             headers={'Accept': 'application/vnd.github.raw'},
         ) as response:
-            if response.status != 200:
+            if response.status != HTTPStatus.OK:
```
Comment (m31-galaxy) [inline breadcord/core_modules/module_manager/__init__.py:90]: Silly question, but is this valid? Does aiohttp do some helpful handling there, or would we have to use `HTTPStatus.OK.value` instead? It makes logical sense but I'm not sure about how thorough you were with testing, so just to be safe

File: breadcord/core_modules/settings_manager/__init__.py [hunk 2]
```diff
@@ -129,7 +130,7 @@ async def autocomplete_value(self, interaction: discord.Interaction, current: st
                 return [app_commands.Choice(name=f"⚠️ Invalid integer '{current}'", value=current)]
 
         elif setting.type == str:
-            if len(current) >= 2 and current[0] == current[-1] and current[0] in '\'"':
+            if current[0] + current[-1] in ('""', "''"):
```
Comment (m31-galaxy) [inline breadcord/core_modules/settings_manager/__init__.py:133]: I'm realising now that this would probably cause an IndexError when `current` is an empty string.

File: breadcord/helpers.py [hunk 3]
```diff
@@ -77,7 +77,7 @@ def search_for(
     threshold: float = 80,
     max_results: int | None = 25,
 ) -> Sequence[_T]:
-    """A custom implementation of a fuzzy search algorithm.
+    """Custom implementation of a fuzzy search algorithm.
```
Comment (m31-galaxy) [inline breadcord/helpers.py]: See previous comment @ L27.

```suggestion
    """A custom implementation of a fuzzy search algorithm.
```

File: breadcord/helpers.py [hunk 4]
```diff
@@ -122,7 +122,7 @@ def simple_button(
     emoji: str | discord.Emoji | discord.PartialEmoji | None = None,
     row: int | None = None,
 ):
-    """A substitute for ``discord.ui.button`` which generates a custom ID for you.
+    """Substitute for ``discord.ui.button`` which generates a custom ID for you.
```
Comment (m31-galaxy) [inline breadcord/helpers.py]: See previous comment @ L27.

```suggestion
    """A substitute for ``discord.ui.button`` which generates a custom ID for you.
```

File: breadcord/helpers.py [hunk 5]
```diff
@@ -165,7 +166,7 @@ def decorator(func: ItemCallbackType[V, Button[V]]) -> ItemCallbackType[V, Butto
 
 
 def simple_transformer(to: type[_T]) -> Callable[[type[_Transformer]], _Transformer]:
-    """A decorator for discord.py transformers which makes them easier to use in type annotations.
+    """Decorator for discord.py transformers to make them easier to use in type annotations.
```
Comment (m31-galaxy) [inline breadcord/helpers.py]: See previous comment @ L27.

```suggestion
    """A decorator for discord.py transformers to make them easier to use in type annotations.
```

File: breadcord/module.py [hunk 6]
```diff
@@ -70,18 +70,17 @@ def load_settings_schema(self) -> None:
 
     def install_requirements(self) -> None:
         distributions = list(importlib.metadata.distributions())
-        missing_requirements = []
-        for requirement in self.manifest.requirements:
+        if missing_requirements := [
```
Comment (m31-galaxy) [inline breadcord/module.py]: guh???

I'm assuming this was an automated linter check, but while this could *technically* be considered better from a robot's perspective, I feel like the complexity of this seriously sacrifices code legibility. Is it worth it?

File: breadcord/core_modules/settings_manager/__init__.py [hunk 7]
```diff
@@ -43,7 +44,7 @@ def __init__(self, bot: breadcord.Bot):
         super().__init__()
 
     async def on_submit(self, interaction: discord.Interaction) -> None:
-        with open(self.bot.settings_file, 'w', encoding='utf-8') as file:
+        async with aiofiles.open(self.bot.settings_file, 'w', encoding='utf-8') as file:
             file.write(self.editor.value)
```
Comment (m31-galaxy) [inline breadcord/core_modules/settings_manager/__init__.py]: If we're switching to `aiofiles` here, would `file.write` not be a coroutine here, and therefore have to be awaited?

```suggestion
            await file.write(self.editor.value)
```

File: pyproject.toml [hunk 8]
```diff
@@ -27,15 +27,23 @@ select = [
 
 ignore = [
     # Missing docstrings
-    # These should be enabled in the future, but for now it's just too much at once
+    # This should be enabled in the future, but for now it's just too much at once
     "D10",
 
+    # I have aged 30 years due to this rule
+    # - Fripe
```
Comment (m31-galaxy) [inline pyproject.toml:34]: Congratulations, I think this is Breadcord's first source code easter egg :D

File: pyproject.toml [hunk 9]
```diff
@@ -27,15 +27,23 @@ select = [
...
+    "D401",
+
     # Wacky docstring defaults
     "D203",
     "D213",
 
     # Insane suggestion that pathlib.Path.open() should be prefered over open()
     "PTH123",
+
+    # Too unflexible to be resonable
```
Comment (m31-galaxy) [inline pyproject.toml]: Typo nitpick

```suggestion
    # Too inflexible to be reasonable
```

File: breadcord/helpers.py [hunk 10]
```diff
@@ -24,7 +24,7 @@
 
 
 async def administrator_check(interaction: discord.Interaction) -> bool:
-    """A discord.py check to ensure that the interaction user is an administrator.
+    """discord.py check to ensure that the interaction user is an administrator.
```
Comment (m31-galaxy) [inline breadcord/helpers.py]: This isn't imperative mood as far as I know, so I'm not sure why this change was made - I preferred it with the `A` left in personally. Any reason for the removal?

```suggestion
    """A discord.py check to ensure that the interaction user is an administrator.
```

File: breadcord/helpers.py [hunk 11]
```diff
@@ -143,10 +143,11 @@ def decorator(func: ItemCallbackType[V, Button[V]]) -> ItemCallbackType[V, Butto
         if not inspect.iscoroutinefunction(func):
             raise TypeError('button function must be a coroutine function')
 
+        max_id_length = 100
         # noinspection PyUnresolvedReferences
         custom_id = f'{func.__module__}:{func.__qualname__}'.removeprefix('breadcord.')
-        if len(custom_id) > 100:
-            raise RuntimeError(f'decorated function path exceeds 100 characters: {custom_id}')
+        if len(custom_id) > max_id_length:
+            raise RuntimeError(f'decorated function path exceeds {max_id_length} characters: {custom_id}')
```
Comment (m31-galaxy) [inline breadcord/helpers.py:150]: Nitpick - constants should be UPPER_SNAKE_CASE

```suggestion
        max_id_length = 100
        # noinspection PyUnresolvedReferences
        custom_id = f'{func.__module__}:{func.__qualname__}'.removeprefix('breadcord.')
        if len(custom_id) > max_id_length:
            raise RuntimeError(f'decorated function path exceeds {max_id_length} characters: {custom_id}')
```

**Review by m31-galaxy — CHANGES_REQUESTED** [review]
I should not be doing this right now but ¯\\\_(ツ)\_/¯

File: breadcord/core_modules/module_manager/__init__.py (hunk 1 above)
Comment (Fripe070) [reply breadcord/core_modules/module_manager/__init__.py:90]:
```py
>>> import http
>>> 200 == http.HTTPStatus.OK
True
```
See the documentation for [http.HTTPStatus](https://docs.python.org/3/library/http.html#http.HTTPStatus), it’s an IntEnum.

File: breadcord/core_modules/settings_manager/__init__.py (hunk 2 above)
Comment (Fripe070) [reply breadcord/core_modules/settings_manager/__init__.py:133]: > I'm realising now that this would probably cause an IndexError when `current` is an empty string.

Yes, but would it ever be empty? Considering that a setting can’t be, the function would return before getting here.

File: breadcord/core_modules/settings_manager/__init__.py (hunk 7 above)
Comment (Fripe070) [reply breadcord/core_modules/settings_manager/__init__.py]: Good catch

File: breadcord/helpers.py (hunk 11 above)
Comment (Fripe070) [reply breadcord/helpers.py:150]: Not inside a function. PyCharm complains.

File: breadcord/core_modules/settings_manager/__init__.py (hunk 2 above)
Comment (m31-galaxy) [reply breadcord/core_modules/settings_manager/__init__.py:133]: I might be a bit dim, but I'm not seeing where it'd return before if it was an empty string

Comment (m31-galaxy) [reply breadcord/core_modules/settings_manager/__init__.py:133]: As discussed on Discord, this should be fine as-is - the code here is awful and needs refactoring anyways, and a bug involving broken autocomplete for empty inputs is the least of our concerns.

**Comment by Sophon96** [comment]
Why not use `response.ok`?

**Comment by Fripe070** [comment]
> Why not use `response.ok`?

That would probably be fine here, but it wouldn’t work the same as the original code ([`aiohttp.ClientResponse.ok`](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientResponse.ok) is `True` if the status code is less than 400, while we currently check if it’s exactly 200)

File: breadcord/module.py [hunk 12]
```diff
@@ -69,17 +69,18 @@ def load_settings_schema(self) -> None:
...
+        def is_installed(requirement: Requirement) -> bool:
+            for distribution in installed_distributions:
+                if requirement.name == distribution.name and distribution.version in requirement.specifier:
+                    return True
+            return False
+
+        if missing_requirements := filter(
+            lambda requirement: not is_installed(requirement),
+            self.manifest.requirements,
+        ):
```
Comment (m31-galaxy) [inline breadcord/module.py]: Why `is_installed` instead of `is_missing`? The latter would be neater and remove the need for a lambda altogether

**Review by m31-galaxy — APPROVED** [review]
LGTM, thanks a lot Fripe for all your hard work on this! You're absolutely carrying the project while I'm away :p
